### DISCUSSION TITLE
Future of framework dependencies

### DISCUSSION THREAD

**Discussion by RossNordby** [bepu/bepuphysics2#80] [General]
.NET Core 3.0 is now [out](https://devblogs.microsoft.com/dotnet/announcing-net-core-3-0/), and it seems like a good time to collect information for the future of bepuphysics v2's dependencies.

At the moment, the library targets .NET Standard 2.0. It can run on .NET Framework, .NET Core, mono, and in theory any other runtime supporting the standard.

However, .NET Standard 2.0 is missing a few features which would be handy to have for performance and simplicity. 2.1 includes some of these features, but requires pretty recent platforms. .NET Framework 4.8, for example, will not support .NET Standard 2.1.

Further, even .NET Standard 2.1 is missing the platform specific intrinsics (at least [as far as I can tell](https://docs.microsoft.com/en-us/dotnet/api/?view=netstandard-2.1)- let me know if I'm wrong about this!). The only full featured path appears to be .NET Core 3.0 (and later, the unified .NET 5).

So, the options:
1. .NET Standard 2.0: compatibility with .NET Framework and current Unity implementation, miss out on mild performance improvements available in new APIs. (The engine already uses `Vector<T>` extensively, so the benefit of platform dependent intrinsics isn't enormous. But it's not zero either.) Notably, runtime bugs in Unity currently block full functionality, and while workarounds can make parts of it work, it tends to be so slow that it's mostly pointless.
2. .NET Standard 2.1: Current .NET Framework versions unsupported, Unity unsupported (until later versions). MathF and other small tweaks become available.
3. .NET Core 3.0: Gain access to platform dependent intrinsics, lose support for every other platform.

The obvious choice is to build for multiple targets, but this is a case where each additional target is enormous, and maintaining legacy paths would be painful. I'd like to see where people are using the engine to see if trimming down the support targets is possible.

Questions for users:
1. What are you targeting now with bepuphysics v2? 
2. Would your project [be compatible with .NET Standard 2.1](https://github.com/dotnet/standard/blob/master/docs/versions.md)? 
3. How about .NET Core 3.0?
4. Has anyone managed to use it within unity? (And if so, could you elaborate on how? :))

(It's worth noting that, even if the decision was made to, say, only focus on .NET Core 3.0, it would be a while before I'd actually work on that, let alone finish it. This is primarily a medium-long term concern.)

**Comment by vpenades** [comment]
As a business, you never know when a client is going to come through the door and tell "hey, that cool project you have on windows? it would be nice if you can run it on Android, or in Windows XP" (which is something we've been actually asked)

Performance is important, but what's the point of a project if it cannot be used? so that answers the question: compatibility goes before performance.

As you say, the ideal solution would be multiple targets, but it's a lot of maintenance for a single developer. It's happening to me too in my projects.

BepuPhysics seems already very efficient and performant, so I would stick with NetStandard2.0.... I don't think it's worth to loose several platforms for the sake of gaining 15% performance.

**Comment by KakCAT** [comment]
I'm not using bepu 2 (and unfortunately maybe I will never use it, as my current project may be my last solo project and I decided to stick with bepu 1) but supposing I would, my requirements would be working nicely with PS4 and Switch and able to work side by side with MonoGame (it does now).

I don't know how this exactly translates as answers to your questions, but I suppose I'm also in the compatibility bandwagon.

**Comment by lidgren** [comment]
.NET Core 3 all the way.

**Comment by Alan-FGR** [comment]
I'd vote for net standard 2.1 atm but since it's something for the long run then yeah net core 3 imho.

**Comment by phende** [comment]
It seems the reason why bepuphysics2 exists and is so different from v1 is about squeezing the best performance out of .Net, and .NET Standard 2.1 does not bring too much in that area. Or does it ?

So what about branching the current state as .NET Standard 2.0 and only applying critical bug fixes there, then driving master toward .NET Core 3 ? Would a compile-time compatibility shim for intrinsics be complicated, for where intrinsics are not supported ?

I don't have older projects to maintain so am obviously biased... :-)

**Comment by Alan-FGR** [comment]
> So what about branching the current state as .NET Standard 2.0 and only applying critical bug fixes there, then driving master toward .NET Core 3 ?

Branching a stable version doesn't sound like a good idea at this point imho.

**Comment by phende** [comment]
> Branching a stable version doesn't sound like a good idea at this point imho.

Right. Another option then is to branch the current to become bepuphysics v2.1 or v3, that achieves the same effect and does not matter much. Only depends on what path @RossNordby likes to follow.

**Comment by RossNordby** [comment]
I should mention that the engine is pretty much guaranteed to, at some point, adopt platform dependent intrinsics for my own purposes. It's just a question of when and what I do about legacy at that point. For example, if I wait long enough that [.NET 5](https://devblogs.microsoft.com/dotnet/introducing-net-5/) becomes the de facto standard and mostly unifies mono and coreclr, there would likely be little point in trying to maintain .NET Framework compatibility. Unity might take a while longer, but... it's effectively unsupported right now anyway.

(Is anyone stuck using the legacy .NET Framework with bepuphysics v2? Could you describe your situation? And do you think that situation will last into late next year with .NET 5?)

>It seems the reason why bepuphysics2 exists and is so different from v1 is about squeezing the best performance out of .Net, and .NET Standard 2.1 does not bring too much in that area. Or does it ?

Not too much (unless I'm wrong about the platform dependent intrinsics being absent). MathF's inclusion does help clean up a few scalar paths a little.

**Comment by RossNordby** [comment]
>Would a compile-time compatibility shim for intrinsics be complicated, for where intrinsics are not supported ?

I have toyed with the idea of a custom tool to do some very simple codegen without some of the nastiness of text templates, but then I have to slap myself because I already have an infinitely long to-do list :)

Depending on the JIT, it may be possible to have zero overhead JIT-time specialization. Things like `Avx.IsSupported` are already JIT intrinsics, so that part is actually pretty easy. The harder part is ensuring there is no wonkiness in the generated assembly caused by things like wrapper structs, various in-theory-free memory casts, or enregistration blocking. RyuJIT has made a lot of progress on this front (most of the core issues may have already been resolved, I haven't actually looked in a few months). If that pans out, then it would be not-absolutely-horrible to have a `ScalarWide` type alongside `Vector3Wide` and friends that abstracts the underlying vector API. It would still require revisiting pretty much every vectorized implementation, but, that's going to happen no matter what.

**Comment by vpenades** [comment]
It is not my personal case, but for example, MonoGame is trying hard to keep compatibility with Net45, because apparently, most console SDKs that support c# are stuck in Net45. And apparently they move forward into adopting newer versions really slowly.

MonoGame has a weird build system just because that, the public repository has the code for most "public" platforms, but then there's a private repository, for which you need to have a console developer ID, that has the code that targets the consoles (closed source under NDAs). So, whenever someone sends a pull request upgrading some framework to a newer version, it's immediately rejected because it might break the consoles code.

Given that BepuPhysics is clearly aiming for videogames, it could be a waste if it blocks itself from consoles. so I would suggest you to ask some developers working with consoles to know more about the console frameworks life cycle so you know what to expect and which frameworks are/will be supported.

**Comment by hawkerm** [comment]
Considering the lack of maintained C# physics engines, bepu looks really promising for adoption on a project I'm working on.

However, I'm planning to not only target UWP, but also WASM using Uno. I know the earlier doesn't support .NET Standard 2.1 yet though.  I'm also not sure how that would effect WASM support either, though I'm not sure if anyone has tried bepu in WASM yet?

**Comment by RossNordby** [comment]
>I'm not sure if anyone has tried bepu in WASM yet?

I haven't seen it- it would be an interesting experiment given the heavy reliance on vectorization. I assume it wouldn't vectorize yet (unless the transpilers and WASM SIMD proposals are further along than I thought), but the scalar fallback may still have acceptable performance if it can handle the compile-time branching properly.

**Comment by Gavin-Williams** [comment]
I'm using .Net Standard libraries, targeting UWP. 

A few people are saying .Net Core 3. There is almost zero reason to target .Net Core. It's not a good desktop platform. By the time it is, it'll be called .Net 5. Microsoft are still working on a lot of the solutions that will unify UWP and Desktop with .Net Core. 

UWP will get .Net Standard 2.1 at some point, not sure when, but I don't think it will be too far away. .Net Standard is the recommended way to write libraries. That will change the day .Net 5 implements .Net Standard, not a day before.

**Comment by impoetk** [comment]
Hi we are currently using it in a WPF application, so we are stuck for a while in .net framework, so .net standard 2.0 is absolutely needed for us

**Comment by RossNordby** [comment]
@impoetk Out of curiosity, have you investigated .NET Core 3's WPF support?

**Comment by reflectronic** [comment]
I think moving over to .NET 5 should be the ultimate goal; .NET Standard will effectively be 'dead' for any new platform development after .NET 5's release. It will also bring various goodies like:

- An AOT compilation scheme, which can help significantly speed up startup time
- `Unsafe.SkipInit`/`SkipLocalsInitAttribute`; lets us address the problem in #59 better
- Support for ARM hardware intrinsics (in addition to the current x86 intrinsics in .NET Core 3)
- Zero-cost conversion between System.Numerics.Vector* types and Hardware Intrinsics types

As an aside, something from .NET Standard 2.1 that I'd like to see is nullable reference types. I experimented with enabling them and actually found some bugs in the library, e.g. places where the code assumed that `default(TSomeGenericType)` was an actual value (and not `null`), even when `TSomeGenericType` is not constrained to `struct`.

**Comment by RossNordby** [comment]
Good catch on the unconstrained default(T) stuff (eb692e70a24409e9af958bf6102709b58b64ccde). In practice, the vast majority of those wouldn't ever become an accessible problem for a user, but a few were potential problems and there's no reason to keep the rest around.

.NET 5 is definitely where I'm leaning at the moment. Considering it's already in preview, if I were to start work, it would definitely involve .NET 5 as opposed to .NET Core 3 or .NET Standard 2.1.

There remains the question of transitioning. To maintain compatibility with all current uses, .NET Standard 2.0 is required. To support both at once requires invasive conditional compilation or equivalent. The good news is that things like the JIT-aware zero cost conversion API reduce the suffering involved quite a bit, but plenty remains.

On the other hand, .NET 5 can, in principle, cover *almost* all the previous use cases (with a few exceptions for older incompatible libraries). 

So, given that it's been a few months, I'd like to repeat some questions to everyone:
1. Is your use case compatible with .NET 5?
2. If not, what is blocking it?
3. If not, for how long do you foresee this incompatibility lasting?

**Comment by lidgren** [comment]
Yes.

**Comment by Gavin-Williams** [comment]
SharpDX isn't compatible with .Net 5 - only .Net Standard AFAIK. I don't know what's going to be done about that. But without a SharpDX style API for C#, I'm not sure there is even a story for C# game/graphics development in .Net 5.

That's one of my greatest concerns right now. There are some graphics API projects under development, but nothing like SharpDX from what I've seen. And that means my engine will stay .Net Standard. And .Net Standard is still the recommended way to build libraries. That hasn't changed.

So.

1. No.
2. .Net Standard provides compatibility with existing, necessary API's.
3. Indefinitely at the moment. There is little support for C# game/graphics at MS.

**Comment by reflectronic** [comment]
> SharpDX isn't compatible with .Net 5 - only .Net Standard AFAIK

.NET 5 will support all .NET Standard libraries, in the same way .NET Core does today.

**Comment by RossNordby** [comment]
(As a side note, it's pretty easy nowadays to autogenerate bindings. That's what I've been doing for private projects. You might want to check out the https://github.com/microsoft/ClangSharp PInvokeGenerator or similar tools. I might eventually release the bindings I've made with it, but it's not hard to do yourself.)

**Comment by Gavin-Williams** [comment]
@reflectronic I don't think it's that simple. 

* Can we use a .Net 5 library (Bepu) inside a .Net Standard library (Engine)? Isn't the compatibility one-way?
* .Net Core doesn't even support desktop scenarios currently - no UWP, no swapchains, no CoreWindow. How do you know that SharpDX is just going to work with API's that don't exist yet? I don't see at all how they will be compatible.

I don't think I'd even try using something like ClangSharp - SharpDX wasn't even able to be supported because of effort involved. So how am I going to generate my own version of SharpDX when an opensource project run by an expert couldn't even maintain it.  And if it was so simple, wouldn't Microsoft be providing C# DirectX? Thanks for the suggestion, but it's hard enough making a game on my own right now, I don't see how I could replace SharpDX in my spare time .. no one else has managed to do it have they? I've seen one or two efforts and they are quirky and require knowledge of C++ and how to use DirectX in that context.

**Comment by RossNordby** [comment]
>Can we use a .Net 5 library (Bepu) inside a .Net Standard library (Engine)? Isn't the compatibility one-way?

You wouldn't be able to directly depend on a .NET 5 project from within a .NET Standard 2.0 project, but SharpDX or other .NET Standard 2.0 third party dependencies do not depend on bepuphysics2. Unless you require the older framework compatibility of .NET Standard for your project, there is not really any reason to use it.

>.Net Core doesn't even support desktop scenarios currently - no UWP, no swapchains, no CoreWindow. How do you know that SharpDX is just going to work with API's that don't exist yet? I don't see at all how they will be compatible.

The bepuphysics demos use SharpDX in .NET Core 2.1. I have private projects on Vulkan/DX12 running with .NET 5 previews.

I'll admit I haven't paid any attention to UWP, since it's looking pretty dead-endy (at least with regard to a "UWP first/only" approach). As far as I know, .NET Core 3 brought with it desktop packs to make it easier to bring UWP/WPF/winforms/EF6/etc applications over. Using a windows-specific pack like WPF will make the application only run on windows, of course, but otherwise it works as expected. I'm not entirely sure if .NET Core 3 is deployable in the store or not- it may be.

My understanding is that .NET 5, at least, should be deployable on the store directly, though I wouldn't recommend taking my word for it.

>I don't think I'd even try using something like ClangSharp

That's fine, you can keep using SharpDX if you'd like, it won't block anything. It really is easier than you might think though :)

**Comment by reflectronic** [comment]
> Can we use a .Net 5 library (Bepu) inside a .Net Standard library (Engine)? Isn't the compatibility one-way?

The idea with .NET 5 is more or less that .NET Standard is going away. All of the platforms that .NET Core and Mono are supporting will come together in .NET 5, so your engine would probably target .NET 5 as well.

> .Net Core doesn't even support desktop scenarios currently - no UWP, no swapchains, no CoreWindow. How do you know that SharpDX is just going to work with API's that don't exist yet? I don't see at all how they will be compatible.

UWP is moving to .NET 5 (running under CoreCLR). Additionally, SharpDX should already work with UWP now, so it will continue to work with UWP in the future.

> I don't think I'd even try using something like ClangSharp

There are already bindings out there, e.g.  https://github.com/terrafx/terrafx.interop.windows (which is also maintained by one of ClangSharp's maintainers). They match almost exactly to the underlying Windows APIs, and will continuously be updated with various .NET features like function pointers etc.

**Comment by Gavin-Williams** [comment]
"The idea with .NET 5 is more or less that .NET Standard is going away" Can you please tell me where that has been said? Or is that just a rumor that community members have been putting forward?

My engine would have to continue to target .Net Standard most likely and any libraries I use would have to be .Net Standard - that provides best compatibility. SharpDX would need to be updated most likely to support .Net 5. I've talked to the developer of SharpDX, even he doesn't know if SharpDX will be usable with .Net 5 because we don't know what it looks like yet, what it's requirements will be.

"UWP is moving to .NET 5 (running under CoreCLR). Additionally, SharpDX should already work with UWP now, so it will continue to work with UWP in the future." - SharpDX works with .Net Standard now as well - that's what's important - are people actually writing UWP libraries? And we don't know what UWP is going to look like under .Net 5 - Won't it just be a container with all C# development moving to .Net 5 ? That would mean the current UWP programming model will go away. We will program in .Net 5 C# and build for the UWP or Win32 containers.

I think it's way too early to be having this discussion. .Net Standard is currently the only way to write libraries that are compatible across frameworks. Both .Net Framework and UWP are still at the table and will be into next year. This is a discussion for 2021/22 not 8 months before the framework even exists. .Net Standard isn't going anywhere any time soon.

**Comment by RossNordby** [comment]
>"The idea with .NET 5 is more or less that .NET Standard is going away" Can you please tell me where that has been said? Or is that just a rumor that community members have been putting forward?

It's not going away in the sense that it's unsupported or no longer usable. There's just less use cases for *new* versions of .NET Standard after unification. As far as I've seen, the discussion around future .NET Standards is of the form 'not sure yet'. For example, from here: https://devblogs.microsoft.com/dotnet/introducing-net-5/
>Do we still need .NET Standard? We are working through these issues now and will soon be sharing design docs for you to read and give feedback on.

The idea is that all new applications can just target .NET 5 and it'll be fully supported on all the available recent runtimes. .NET Standard is only useful when trying to offer a common denominator across APIs of different capabilities, like .NET Framework and .NET Core. .NET Standard 2.0, for example, is already supported by recent version of both of those.

.NET 5, critically, is a superset of .NET Core. It implements .NET Standard 2.0. I can flip the switch in the bepuphysics demos to netcoreapp5.0 right now with SharpDX as a .NET Standard 2.0 dependency and everything works perfectly. This is not a thought experiment- I actually did it. SharpDX can be used from a .NET 5 application.

Now, would SharpDX be able to *directly* target .NET 5 without some changes? Dunno, bepuphysics2 can't yet because of the stuff I did with the reference dummies, but that doesn't matter for the consumer of a .NET Standard 2.0 SharpDX. There's no need to update SharpDX to directly target .NET 5.

>I think it's way too early to be having this discussion. .Net Standard is currently the only way to write libraries that are compatible across frameworks.

It's true that .NET Standard 2.0 is required to have a single target across old frameworks. The purpose of this issue is to figure out how many users of bepuphysics2 are targeting those old frameworks, and if they foresee still being in that situation after .NET 5 releases. That's valuable information for me even if I'm not currently in the process of pushing bepuphysics2 over.

**Comment by nick-dodonov** [comment]
We are using Bepuphysics2 in Unity to simplify development (when the server physics processing is integrated into the build).

* No need for special export-pipeline or external process execution (fast run iterations - very useful for artists).
* It also allows to render and debug an internal physics state easily. 

So the loss of .netstandart2.0 is dramatic for us.

**Comment by RossNordby** [comment]
Are you successfully running bepuphysics2 in unity with decent performance? That would be pretty interesting- last I heard (months ago) there were still some orders of magnitude slowdowns.

**Comment by hawkerm** [comment]
This seems like a timely article from the .NET Team: [The Future of .NET Standard](https://twitter.com/terrajobst/status/1306021755752861697?s=20)

Still, if you want folks using .NET Framework, Unity, and UWP to be able to use the library into 2021, you should stick to targeting .NET Standard 2.0. There are many libraries/APIs available from 2.1 that are available in NuGet packages which work for 2.0 as well still.

**Comment by hawkerm** [comment]
@Sergio0694 would any of your new [High Performance APIs](https://docs.microsoft.com/en-us/windows/communitytoolkit/high-performance/introduction) be helpful here? As you already do a lot of the abstraction to multi-target both 2.0 and 2.1 for where 2.1 can be taken advantage of, right?

That could allow the benefits of gaining some performance while maintaining compatibility, without the extra overhead on the bepu repo to worry about the multi-targeting aspects that goes along with that?

**Comment by Sergio0694** [comment]
@hawkerm Yes, the `HighPerformance` package does abstract away a lot of runtime specific details to offer platform specific optimizations in many of the available extensions (eg. to avoid bounds checks when indexing arrays on .NET Core 2.0+, or using intrinsics on .NET Core 3.1 for the `BitHelper` APIs, etc.). Those would definitely be usable, but for `bepuphysics2` in particular I think the actual performance bottleneck would just be domain specific SIMD pipelines, as mentioned in the first post. And that's not something we offer (or that we could offer), as it really just depends on what you're trying to do exactly.

@RossNordby My general suggestion here would probably be to do what `ImageSharp` did as well, ie. to multi-target .NET Standard (in this case, .NET Standard 2.0), and .NET Core 3.1 (since it's the LTS release and also the one with intrinsics).
You could keep all your codebase practically as is, with the library still working just fine on both .NET Framework, UWP and Unity, and with `Vector<T>` doing all the SIMD work whenever possible. On .NET Core instead, you would be able to rewrite your SIMD pipelines with the actual intrinsics APIs to get the best performance. This approach has really been working well with `ImageSharp`, especially because it makes things very easy to use for consumers as well.

Imagine you write a .NET Standard 2.0 library referencing `ImageSharp`. If your library is then referenced by a UWP app, NuGet will resolve the .NET Standard 2.0 as expected, while a .NET Core 3.1 consumer will still end up resolving the .NET Core 3.1 target, even if it's an indirect dependency. So you really get the best of both worlds - maximum reach and still very good performance. The only drawback is having some `#idfef`-s here and there, but nothing that proper code architecture can't hide away to make life easier for library maintainers. Optimized code is often not the prettiest code, but then it's also the prettiest because it's optimized 😄

**Comment by RossNordby** [comment]
Unfortunately, the divergence in optimized implementation can be sufficiently huge between .NET Core 2.0/Framework and .NET 5 that I realistically can't/won't keep both at locally maximum performance and feature parity in the long term. Beyond API differences, there are improvements in the JIT that enable large scale structural simplifications- but that simpler implementation would cause older runtimes to choke badly.

Making large new changes that maximize performance while keeping old framework/runtime compatibility means doing about three times more work, which in practice means I'll make fewer changes. That's already happened- I've avoided making certain big changes because they benefit so much from .NET 5 and creating an equivalent legacy path would be a huge pain.

**What I'd really like to know is how many users are actually using bepuphysics2 on one of these older targets, and how many foresee still being in that situation as .NET 5 and .NET 6 roll out.** 

In particular, I'd be very interested in hearing from anyone who's managed to actually get bepuphysics2 working well with Unity (@nick-dodonov ;)) since the last reports I heard suggested that it was effectively unusable due to either compiler problems or >100x slowdowns.

In other words, I'd like to identify the non-hypothetical users who would be troubled by a future transition before putting in hundreds of hours of extra work :P

**Comment by Bartolomeus-649** [comment]
My take on this...

**In the long run this is not an issue**
* .NET Framework won't get any new features from Microsoft since they have clearly said that future versions of  .NET Core (i.e. .NET 5, .NET 6, ...) will fill most gaps that (not WCF for example) exists today between .NET Framework and .NET Core. And since .NET Framework is "closed source", no one else can add anything to the BCL.
* Practically all new 3:ed party libraries and packages for the ".NET Universe" are already targeting, or soon will be, targeting .NET Core.
* The whole point of .NET Standard was to provide a method for developers to target an interface. This was needed since there are now several implementations of the .NET runtime environment, and each of them implement different parts of the BCL. So by creating .NET Standard and defining the interface which needs to be implemented for a specific level of the .NET standard, a developer could target the standard and know it would work in any environment where the .NET runtime would support the same .NET standard.
    Going forward, this won't be the case. Microsoft's ambition is to provide one .NET wich can be used and runs everywhere. Everything will be consolidated into one .NET implementation, so no more mono, no more .NET Framework and so on. So by targeting what we today call .NET Core the code will run everywhere.
   The .NET Standard is not going away, it's just that going forward, the only "line"in the .NET Standard matrix which will support future .NET Standard versions will be .NET Core. So,.NET core will, defacto, be the standard.

So, I think the real issue here is what to do during the time it takes for everyone to get on the .NET core train?
For those that already use .NET Core (or can switch to .NET core without any problems) there's no issue.

So, basically we are left with those who use .NET framework and rely, directly or indirectly via dependant libraries, on features/functionality which won't be ported to .NET 5.

Without knowing the specific implementation details of bepuphysics2, but imagining that core issue codewise is to be able to utilize the framework and actual runtime platform (device, hardware, OS, ...) in the best way possible for maximum performance.  Based on this I can think of these solutions which would support both .NET Framework, .NET Core (.NET5, ...) and the device specific implementations of .NET (Super sets of.NET with added API:s and optimized JIT for the specific devices).

* Create a Hardware Abstraction Layer (HAL) which provides conceptual, high level interfaces for the performance critical operations. Then create a HAL implementation for each platform/device.

* Create a [Domain Specific Language](https://docs.microsoft.com/en-us/visualstudio/modeling/modeling-sdk-for-visual-studio-domain-specific-languages?view=vs-2019) (DSL) which would allow for declarative specification of the performance critical operation. These declarations would then be transformed to device/framework specific C#/IL code at compile time.

* Create an interface for a "calculation module". Then create implementations of the interface or each framework/device.


If optimizing for performance, the DSL is probably the best approach. It could be implemented without any overhead, and could be implemented at different levels in the architecture, like, everything from adding two values together to more complex operations like transform a matrix or check if anything has collided with anything else.
This would also allow for things like different data types on different devices. Also, it's possible to start small and let the DSL grow over time. Also, the DSL can be gradually introduced into the library, there's no need to replace every existing implementation of something just because there might be DSL support for it. 
The overhead here is to create/implement the actual DSL and decide how you want to describe the operations you want the DSL to support, and figure out how best integrate it with what exists today. And then create the actual DSL implementation extension for Visual Studio.

**Comment by impoetk** [comment]
> @impoetk Out of curiosity, have you investigated .NET Core 3's WPF support?

not yet, the application is quite complex and i think it will take a long time before we move out of the .net framework, my guess would be to wait for .net 5 and slowly make some tests what will run and what will not, also we depend a lot in 3rd party libraries specific (afaik) to .net framework wpf implementation

**Comment by Gavin-Williams** [comment]
"Practically all new 3:ed party libraries and packages for the ".NET Universe" are already targeting, or soon will be, targeting .NET Core." - that's not true at all. Practically all 3rd party libraries are targeting .net standard. .net core hasn't even had support for desktop until very recently. And .net core is not .net 5. Desktop developers won't really have to consider .net core at all. I would say most .Net Framework and UWP developers will go straight to .Net 5 after it is feature complete - from 2021/22 onward.

.net core has never been a target for libraries (except maybe for those people that jumped into .net core).

![chart-all](https://user-images.githubusercontent.com/4374176/95520081-3fbd6d00-0a12-11eb-99ac-a42cf0e48135.png)

**Comment by RossNordby** [comment]
Given continued progress in runtime unification and adoption, 2.4.0-beta5 now targets .NET 5.0. The intent is to have 2.3.0 be the last version targeting .NET Standard 2.0.

While I cannot promise to change any major plans in response, I would still be interested in hearing from any users for whom this is a problem.
